### PR TITLE
feat: clean urls module

### PR DESCRIPTION
**Pull request by Rapougnac**
This module allows members to send any link, and be automatically cleaned/trimmed if the bot detects undesirable query parameters.

Here's a quick demo:
out-2024-07-15_21 56 53

The processing is faster, it's just that my internet is damn slow

### CODE REVIEW

**Comment by SirLynix** [comment]
tell me when you're done for a review!

**Comment by Rapougnac** [comment]
FWIW, this PR is ready, it's just still missing to get the full URL of a shortened link

File: module_clean_urls.lua [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+    ["twitter.com"] = "vxtwitter.com",
+    ["x.com"] = "fixvx.com",
+}
+
+Module.UsersHanging = {}
+
+Module.UniversalRules = {}
+Module.HostRules = {}
+Module.RulesByHost = {}
+Module.GuildUniversalRules = {}
```
Comment (SirLynix) [inline module_clean_urls.lua:249]: Why not use guild persistent data?

File: module_clean_urls.lua [hunk 2]
```diff
@@ -0,0 +1,737 @@
...
+    self.GuildRulesByHost[guildId] = nil
+end
+
+local function removeParam(rule, param, queryParams)
+    if param == rule or string.match(param, rule) then
+        queryParams[param] = nil
+    end
+end
+
+local function request(method, url)
```
Comment (SirLynix) [inline module_clean_urls.lua:335]: what's the point vs coro http.request?

File: module_clean_urls.lua [hunk 3]
```diff
@@ -0,0 +1,737 @@
...
+end
+
+--- Find the location of a URL
+---@param url string The URL to resolve
+---@return string|nil ... The resolved location, or the original URL if no location was found
+local function resolveLocation(url)
+    if not url then return end
+
+    ---@type table
+    local headers, _ = request("HEAD", url)
```
Comment (SirLynix) [inline module_clean_urls.lua:357]: can't you do that using a single GET request and inspect the headers?

File: module_clean_urls.lua [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+
+    self:CreateGuildRules(config, guild.id)
+    self:SaveGuildConfig(guild)
+end
+
+function Module:OnLoaded()
+    self:CreateRules()
+
+    self:RegisterCommand({
+        Name = "clean",
```
Comment (SirLynix) [inline module_clean_urls.lua:478]: clean is a very generic command name, maybe cleanurl to avoid confusion?

File: module_clean_urls.lua [hunk 5]
```diff
@@ -0,0 +1,737 @@
...
+            username = message.author.globalName or message.author.username,
+            content = replaced,
+            components = components
+        },
+        { wait = true }
+    )
+
+    self.UsersHanging[message.author.id] = true
+
+    setTimeout(config.ButtonTimeout or 10000, function()
```
Comment (SirLynix) [inline module_clean_urls.lua:666]: You can use the bot timers instead

File: module_clean_urls.lua (hunk 2 above)
Comment (Rapougnac) [reply module_clean_urls.lua:335]: Accidental push while testing some things, will remove it

File: module_clean_urls.lua (hunk 1 above)
Comment (Rapougnac) [reply module_clean_urls.lua:247]: How does it work?

Comment (SirLynix) [reply module_clean_urls.lua:246]: You can use `self:GetPersistentData(guild)` to get a permanent (saved) per-guild table, and `self:GetGetData(guild)` to get temporary (wiped on bot reboot, ideal for caching) per-guild table.

Comment (Rapougnac) [reply module_clean_urls.lua:246]: Sweet, ty, will implement this tomorrow

**Pull request merged**